pkg/operator/resource/resourceapply/core.go

```diff
@@ -3,42 +3,102 @@ package resourceapply
 import (
 	"bytes"
 	"context"
+	"crypto/md5"
 	"fmt"
+	"io"
 	"sort"
 	"strings"
 
-	"k8s.io/klog/v2"
-
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
 	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog/v2"
 
 	"github.com/openshift/library-go/pkg/operator/events"
 	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
 )
 
+type CachedVersionKey struct {
+	name, kind, namespace string
+}
+
+type CachedResource struct {
+	resourceHash, resourceVersion string
+}
+
+// map of name,kind to resourceHash (of last applied resource) : resourceVersion(when resource was applied)
+var cachedResourceVersions map[CachedVersionKey]CachedResource = make(map[CachedVersionKey]CachedResource)
+
+func updateCachedResourceVersion(name string, kind string, namespace string, resourceHash string, resourceVersion string) {
+	if resourceVersion == "" {
+		klog.V(4).Infof("resourceVersion has empty string %s:%s:%s %s", name, kind, namespace, resourceVersion)
+		return
+	}
+	klog.V(7).Infof("updated resourceVersion of %s:%s:%s %s", name, kind, namespace, resourceVersion)
+	cachedResourceVersions[CachedVersionKey{name, kind, namespace}] = CachedResource{resourceHash, resourceVersion}
+}
+
+func clearCachedResourceVersion() {
+	cachedResourceVersions = make(map[CachedVersionKey]CachedResource)
+}
+
+// if the manifest applied is the same as the last time this resource was
+// updated, and the resourceVersion also hasn't changed, we don't need to
+// actually do anything
+func checkCachedResourceVersion(name string, kind string, namespace string, resourceHash string, resourceVersion string) bool {
+	var versionMatch, hashMatch bool
+	if cached, exists := cachedResourceVersions[CachedVersionKey{name, kind, namespace}]; exists {
+		versionMatch = cached.resourceVersion == resourceVersion
+		hashMatch = cached.resourceHash == resourceHash
+		if versionMatch && hashMatch {
+			klog.V(7).Infof("found matching resourceVersion & manifest hash")
+			return true
+		}
+	}
+	return false
+}
+
+func hashOfResourceStruct(o interface{}) string {
```

Comment:
please note that this only works if the string representation has changed.  Not all changes result in changed ToString functions.  But it's a pretty narrow edge.

Reply:
I think with the reduced scoped of the last refactor, that this won't come up. I couldn't get find this edge case in my testing - under what circumstances would the string representation of a Resource from the apiserver not change when the Resource has changed?

Reply:
> I think with the reduced scoped of the last refactor, that this won't come up. I couldn't get find this edge case in my testing - under what circumstances would the string representation of a Resource from the apiserver not change when the Resource has changed?

nil versus empty is pretty classic.  It has come up a few times when handling fuzzing tests.  I'm ok with the gap as long as it is described.

Reply:
Added description